Add mulitipathd_info text collector example

ssinha-ionos opened the pull request:
multipathd_info is a script that exposes device mapper multipathing
metrics from multipathd daemon.

Signed-off-by: Saket Sinha <saket.sinha@cloud.ionos.com>

SuperQ commented on text_collector_examples/multipathd_info:
```diff
@@ -0,0 +1,9 @@
+#/bin/bash
```
Shouldn't this be `#!/bin/bash`?

ssinha-ionos replied:
This doesn't need to be bash. There is nothing in this script that uses bash-specific features, so it should simply be /bin/sh.

dswarbrick replied:
Shouldn't it be `#!/bin/sh`? ;-)

I don't see any bash-specific code in the script.

SuperQ commented on text_collector_examples/multipathd_info:
```diff
@@ -0,0 +1,9 @@
+
+echo '# HELP node_dmpath_info State info for dev-mapper path'
+echo '# TYPE node_dmpath_info gauge'
+/sbin/multipathd show paths | /usr/bin/awk '{ if ( NR > 1  ) {print "node_dmpath_info{device=\""$2"\"," "path_state=\""$5"\"," "dm_path_state=\""$6"\"}" " 1"}}'
```
I'm not very familiar with the output of this, but it seems like sometimes the output makes the columns squashed together.

If you look at [this](https://access.redhat.com/solutions/17303), the columns are squished together for `[active]`.

ssinha-ionos replied:
Sorry, I cant access [this](https://access.redhat.com/solutions/17303).

On my system the output looks like something below -

ssinha@ps101a-102:~ sudo multipathd show paths
hcil      dev  dev_t   pri dm_st  chk_st dev_st  next_check      
4:0:0:1   sdc  8:32    1   active ready  running X......... 7/40 
5:0:0:1   sde  8:64    1   active ready  running XX........ 11/40
8:0:0:2   sdl  8:176   1   active ready  running XXXXXXXXX. 36/40

ssinha@ps101a-102:~ sudo /sbin/multipathd show paths | /usr/bin/awk '{ if ( NR > 1  ) {print "node_dmpath_info{device=\""$2"\"," "path_state=\""$5"\"," "dm_path_state=\""$6"\"}" " 1"}}'
node_dmpath_info{device="sdc",path_state="active",dm_path_state="ready"} 1
node_dmpath_info{device="sde",path_state="active",dm_path_state="ready"} 1
node_dmpath_info{device="sdl",path_state="active",dm_path_state="ready"} 1

Let me know if you find something wrong with this.

SuperQ replied:
Ok, I guess that version has different output compared to the other one. Seems OK.

dswarbrick commented:
It might be prudent to specify the exact format that we want `multipathd` to output, in case the order of the columns changes in future (thus breaking the awk script).

Based on the columns selected by the awk script, we want dev, dm_st and chk_st, which are %d, %t and %T respectively (check `multipathd show wildcards`).

This produces a cleaner output which is safer (and probably easier) to parse:
```
$ sudo multipathd show paths format '%d %t %T'
dev dm_st  chk_st
sdd active ready 
sdb active ready 
sdh active ready 
sdf active ready
```

SuperQ commented:
@dswarbrick Nice suggestion. It's too bad that the format string is picky, I get errors when I try and format the whole metric line with it. :frowning_face:

ssinha-ionos commented:
> @dswarbrick Nice suggestion. It's too bad that the format string is picky, I get errors when I try and format the whole metric line with it.

Modified the commit based on @dswarbrick review comments.

Also corrected the metric name -
1. dm_st refers to the DM module(kernel)'s view of the path's state. It can be in one of the two states: 
- active: Last I/O sent to this path successfully completed. Analogous to ready path state.
- failed: Last I/O to this path failed. Analogous to faulty path state.

2. chk_st refers to the physical state of a path. A path can be in one of the following states:
- ready: Path is up and can handle I/O requests.
- faulty: Path is down and cannot handle I/O requests.
- ghost: Path is a passive path. This state is shown in the passive path in Active/Passive mode.
- shaky: Path is up, but temporarily not available for I/O requests.

dswarbrick commented:
Shebang is still incorrect (see @SuperQ's earlier comment):
```
$ shellcheck multipathd_info 

In multipathd_info line 1:
#/bin/bash
 ^-- SC1113: Use #!, not just #, for the shebang.
```

bdrung commented:
And please use `/bin/sh` instead of `/bin/bash` since you do not use bash specific feautures.

discordianfish commented:
LGTM
